### ISSUE TITLE
[Feature] DBT grants should skip grants on managed schemas

### ISSUE THREAD

**Issue by sfc-gh-kumaurya** [dbt-labs/dbt-snowflake#527]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-snowflake functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

This is the feature related to DBT native grants.
 DBT grants feature should be able to support:

1. Skipping grants on exceptional "managed" schemas
2. Skipping grants on roles in environments where they do not exist

### Describe alternatives you've considered

Could be similar to "generate_schema_name" we can have "generate_role_name" that aborts the grants when it returns "null"

**Comment by dbeatty10** [comment]
Thanks for reaching out @sfc-gh-kumaurya !

These are interesting ideas to manage the complexity of grants across different environments.

How were you thinking of specifying the list of exceptional "managed" schemas to exclude? How would you handle this situation today with the current behavior?

Your alternative idea for a special-purpose `generate_role_name` macro is interesting -- could you share more about the details of how it might work? e.g., what would be the inputs and outputs, etc?

**Comment by sfc-gh-kumaurya** [comment]
Hi @dbeatty10,
Thanks for picking up this issue.
For the first question regarding managed schemas, you can go through this Snowflake documentation on "[managed schemas](https://docs.snowflake.com/en/user-guide/security-access-control-configure#creating-managed-access-schemas)" and this medium article on "[Managed schema in snowflake](https://medium.com/@tandon559/managed-schema-in-snowflake-ab7962fd3663)" for example. 
In short, If the schema is regular schema then the owner of an object can grant further privileges on their objects to other roles but in the case of managed schemas, only the schema owner has the power to grant privileges to the objects in that schema (i.e the object owner won't be able to grant further privileges to other roles).

**Comment by sfc-gh-kumaurya** [comment]
Regarding the second question about handling this situation within Snowflake, we've developed a basic macro to manage it. This macro operates by accepting two lists, one containing the permitted roles and another containing the managed schemas. It will only carry out the "grant select to roles" query if the role is included in the allowed list and the schema is not a managed schema.
We want DBT grants to support that feature.

```
{% macro grant_select_to_roles(model) %}

    {% set tag_value = model.config.meta.get('grant_select_to_roles') %}
    {% set allow_list = var('grant_select_to_roles').allowed_roles[target.name] or [] %}
    {% set managed_schemas_list = var('grant_select_to_roles').managed_schemas[target.name] or [] %}

    {% if tag_value is not none %}
        {% for role in tag_value %}
            {# if role is in the allowed list of roles or if an allow list is not defined (deny-by-default) #}
            {% if role in allow_list and model.schema not in managed_schemas_list %}
                {% call statement('create_grant', fetch_result=True) -%}
GRANT SELECT ON {{ this }} TO ROLE {{ role }}
                {%- endcall %}
            {%- endif -%}
        {% endfor %}
    {% endif %}
{% endmacro %}

```

**Comment by dbeatty10** [comment]
Thanks for those links and the example macro @sfc-gh-kumaurya !

I can see now the crucial difference in the table [here](https://docs.snowflake.com/en/user-guide/security-access-control-configure#creating-managed-access-schemas) when managed access is `enable`d or `disable`d:

| Role                                      | Can grant object privileges in a regular schema | Can grant object privileges in a managed access schema |
|-------------------------------------------|-------------------------------------------------|--------------------------------------------------------|
| Schema owner                              | ❌                                              | ✅                                                    |
| Object owner                              | ✅                                             | ❌                                                     |

Is there a way to query Snowflake to determine the `MANAGED ACCESS` configuration on a per schema basis?

i.e., is there some query that can get results similar to this?

| schema_catalog | schema_name         | managed_access |
|----------------|---------------------|----------------|
| prod           | my_regular_schema | disabled        |
| prod           | my_managed_schema   | enabled         |

**Comment by sfc-gh-kumaurya** [comment]
Hi,
Sorry for the late response.
Running the "[Show Schemas](https://docs.snowflake.com/en/sql-reference/sql/show-schemas)" query will give you the desired result. You can see which schema is managed and which is not under the "options" column.


created_on | name | owner | ... | options
-- | -- | -- | -- | --
xyz | my_regular_schema | owner1 | ... | 
xyz | my_managed_schema | owner2 | ... | MANAGED ACCESS
